Slim/Control/Queries.pm

```diff
@@ -752,71 +752,50 @@ sub albumsQuery {
+					SELECT contributor_track.role AS role, contributors.name AS name, contributors.id AS id
+					FROM tracks
+					JOIN contributor_track ON tracks.id = contributor_track.track
+					JOIN contributors ON contributors.id = contributor_track.contributor
+					WHERE tracks.album = :album AND tracks.work = :work
+						AND ( (:performance IS NULL AND tracks.performance IS NULL) OR tracks.performance = :performance )
+						AND contributor_track.role IN (%s)
+					GROUP BY contributor_track.role, contributors.name, contributors.id
+					ORDER BY contributor_track.role, contributors.namesort
 				},
-				join(',', map { Slim::Schema::Contributor->typeToRole($_) } @roles));
+				join(',', map { Slim::Schema::Contributor->typeToRole($_) } @linkRoles));
 			} else {
-				my @roles = ( 'ARTIST', 'ALBUMARTIST' );
+				@linkRoles = ( 'ARTIST', 'ALBUMARTIST' );
 
 				if ($prefs->get('useUnifiedArtistsList')) {
-					# Loop through each pref to see if the user wants to show that contributor role.
-					foreach (Slim::Schema::Contributor->contributorRoles) {
-						if ($prefs->get(lc($_) . 'InArtists')) {
-							push @roles, $_;
+					my %roleMap = %{Slim::Schema::Contributor::roleToContributorMap()};
+					# Loop through roles in role number sequence to see if the user wants to show that contributor role.
+					foreach my $role (sort {$a <=> $b} keys %roleMap) {
+						if ($prefs->get(lc($roleMap{$role}) . 'InArtists')) {
+							push @linkRoles, $roleMap{$role};
 						}
 					}
 				}
-
-				$contributorSql = sprintf( qq{
-					SELECT GROUP_CONCAT(contributors.name, ',') AS name, GROUP_CONCAT(contributors.id, ',') AS id
-					FROM contributor_album
-					JOIN contributors ON contributors.id = contributor_album.contributor
-					WHERE contributor_album.album = :album AND contributor_album.role IN (%s)
-					GROUP BY contributor_album.role
-					ORDER BY contributor_album.role DESC
-				}, join(',', map { Slim::Schema::Contributor->typeToRole($_) } @roles) );
-
 				# when filtering by role, put that role at the head of the list if it wasn't in there yet
 				if ($roleID) {
-					unshift @roles, map { Slim::Schema::Contributor->roleToType($_) || $_ } split(/,/, $roleID);
+					unshift @linkRoles, map { Slim::Schema::Contributor->roleToType($_) || $_ } split(/,/, $roleID);
 					my %seen;
-					@roles = reverse grep !($seen{$_}++), reverse @roles;
-
-					$contributorSql = sprintf( qq{
-						SELECT GROUP_CONCAT(c.name, ',') AS name, GROUP_CONCAT(c.id, ',') AS id
-						FROM (
-							SELECT contributors.name AS name, contributors.id AS id
-							FROM contributor_album
-								JOIN	contributors ON contributors.id = contributor_album.contributor
-							WHERE contributor_album.album = :album AND contributor_album.role IN (%s)
-							GROUP BY contributors.id
-							ORDER BY contributor_album.role DESC
-						)
-						AS c;
-					}, join(',', map { Slim::Schema::Contributor->typeToRole($_) } @roles) );
+					@linkRoles = grep !($seen{$_}++), @linkRoles;
```

Comment:
`reverse` was wrong: it was removing the value just added to the front of the array rather than the duplicate that was already there. This ensures that the specific role passed in as a parameter is placed first. The previous behaviour didn't matter much, because the SQL was doing the sequencing, but now the correct sequence of elements in `@linkRoles` is essential.

Reply:
Wouldn't using the numerical role IDs rather than the string version simplify this? The sort order then would be given.